source/opt/tree_iterator.h

```diff
@@ -115,6 +115,131 @@ class TreeDFIterator {
   std::stack<std::pair<NodePtr, NodeIterator>> parent_iterators_;
 };
 
+// Helper class to iterate over a tree in a depth first post-order.
+// The class assumes the data structure is a tree, tree node type implements a
+// forward iterator.
+// At each step, the iterator holds the pointer to the current node and state of
+// the walk.
+// The state is recorded by stacking the iteration position of the node
+// children. To move to the next node, the iterator:
+//  - Looks at the top of the stack;
+//  - If the children iterator has reach the end, then the node become the
+//    current one and we pop the stack;
+//  - Otherwise, we save the child and increment the iterator;
+//  - We walk the child sub-tree until we find a leaf, stacking all non-leaves
+//    states (pair of node pointer and child iterator) as we walk it.
+template <typename NodeTy>
+class PostOrderTreeDFIterator {
+  static_assert(!std::is_pointer<NodeTy>::value &&
+                    !std::is_reference<NodeTy>::value,
+                "NodeTy should be a class");
+  // Type alias to keep track of the const qualifier.
+  using NodeIterator =
+      typename std::conditional<std::is_const<NodeTy>::value,
+                                typename NodeTy::const_iterator,
+                                typename NodeTy::iterator>::type;
+
+  // Type alias to keep track of the const qualifier.
+  using NodePtr = NodeTy*;
+
+ public:
+  // Standard iterator interface.
+  using reference = NodeTy&;
+  using value_type = NodeTy;
+
+  static inline PostOrderTreeDFIterator begin(NodePtr top_node) {
+    return PostOrderTreeDFIterator(top_node);
+  }
+
+  static inline PostOrderTreeDFIterator end(NodePtr sentinel_node) {
+    return PostOrderTreeDFIterator(sentinel_node, false);
+  }
+
+  bool operator==(const PostOrderTreeDFIterator& x) const {
+    return current_ == x.current_;
+  }
+
+  bool operator!=(const PostOrderTreeDFIterator& x) const {
+    return !(*this == x);
+  }
+
+  reference operator*() const { return *current_; }
+
+  NodePtr operator->() const { return current_; }
+
+  PostOrderTreeDFIterator& operator++() {
+    MoveToNextNode();
+    return *this;
+  }
+
+  PostOrderTreeDFIterator operator++(int) {
+    PostOrderTreeDFIterator tmp = *this;
+    ++*this;
+    return tmp;
+  }
+
+ private:
+  explicit inline PostOrderTreeDFIterator(NodePtr top_node)
+      : current_(top_node) {
+    if (current_) WalkToLeaf();
+  }
+
+  // end() iterator.
```

Comment:
I don't understand this part of the comment.  This is a constructor.

Reply:
That's the Ctor for the end() iterator, I fixed that